File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def doValid(root, lo, hi):
            if root is None:
                return True
            
            if lo is not None and lo >= root.val:
                return False

            if hi is not None and hi <= root.val:
                return False
            
            l = doValid(root.left, lo, root.val)
            r = doValid(root.right, root.val, hi)

            return l and r
        
        return doValid(root, None, None)

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ret{};
        queue<pair<TreeNode*, int>> next{};
        if (root) next.push({root, 0});
        while (!next.empty()) {
            const auto top{next.front()};
            next.pop();
            if (top.second + 1 > ret.size()) ret.push_back({});
            ret[top.second].push_back(top.first->val);
            if (top.first->left) next.push({top.first->left, top.second + 1});
            if (top.first->right) next.push({top.first->right, top.second + 1});
        }
        return ret;
    }
};
